From 1b8bf8ccb58c131efcbb3925bcbb5bd4c14adc03 Mon Sep 17 00:00:00 2001
From: Artem Pelenitsyn <a.pelenitsyn@gmail.com>
Date: Fri, 8 Jul 2022 09:25:30 -0400
Subject: [PATCH] cabal init -i should autodetect author name and maintainer
 email (#8267)

* cabal init -i should autodetect author name and maintainer email (fix #8255)

Name and email are fetched from git config in the non-interactive init
already, and we plug that utility into the interactive one in this commit.

Testing becomes clunkier: the list of inputs have to hold an additional
string that will be used as the return value of
`Interactive.readProcessWithExitCode`, and here is why. Recall that this
method is used to call `git config` to perform autodetection. In the
real scenario (the `IO` instance of `Interactive`), it actually calls
out to `git`. In the testing scenario (the `PurePrompt` instance of
`Interactive`), the method simply returns the next element of the list
of pre-defined inputs. So, the list of inputs to every relevant test has
to be adjusted.

* Add test checking that username/email can be autodetected
---
 .../Client/Init/Interactive/Command.hs        | 11 +++++---
 .../Distribution/Client/Init/Golden.hs        |  2 ++
 .../Distribution/Client/Init/Interactive.hs   | 28 ++++++++++++++++---
 changelog.d/issue-8255                        |  4 +++
 4 files changed, 37 insertions(+), 8 deletions(-)
 create mode 100644 changelog.d/issue-8255

diff --git a/cabal-install/tests/UnitTests/Distribution/Client/Init/Golden.hs b/cabal-install/tests/UnitTests/Distribution/Client/Init/Golden.hs
index 4e02aa3c7..789f110c6 100644
--- a/cabal-install/tests/UnitTests/Distribution/Client/Init/Golden.hs
+++ b/cabal-install/tests/UnitTests/Distribution/Client/Init/Golden.hs
@@ -340,7 +340,9 @@ pkgArgs = fromList
     , "y"
     , "0.1.0.0"
     , "2"
+    , "git username"
     , "foo-kmett"
+    , "git email"
     , "foo-kmett@kmett.kmett"
     , "home"
     , "synopsis"
diff --git a/cabal-install/tests/UnitTests/Distribution/Client/Init/Interactive.hs b/cabal-install/tests/UnitTests/Distribution/Client/Init/Interactive.hs
index 9a998e1f3..f8393d55a 100644
--- a/cabal-install/tests/UnitTests/Distribution/Client/Init/Interactive.hs
+++ b/cabal-install/tests/UnitTests/Distribution/Client/Init/Interactive.hs
@@ -145,8 +145,10 @@ createProjectTest pkgIx srcDb = testGroup "createProject tests"
               -- license
               , "3"
               -- author
+              , "git username"
               , "Foobar"
               -- email
+              , "git email"
               , "foobar@qux.com"
               -- homepage
               , "qux.com"
@@ -249,8 +251,10 @@ createProjectTest pkgIx srcDb = testGroup "createProject tests"
               -- license
               , "3"
               -- author
+              , "git username"
               , "Foobar"
               -- email
+              , "git email"
               , "foobar@qux.com"
               -- homepage
               , "qux.com"
@@ -338,8 +342,10 @@ createProjectTest pkgIx srcDb = testGroup "createProject tests"
               -- license
               , "3"
               -- author
+              , "git username"
               , "Foobar"
               -- email
+              , "git email"
               , "foobar@qux.com"
               -- homepage
               , "qux.com"
@@ -414,8 +420,10 @@ createProjectTest pkgIx srcDb = testGroup "createProject tests"
               -- license
               , "3"
               -- author
+              , "git username"
               , "Foobar"
               -- email
+              , "git email"
               , "foobar@qux.com"
               -- homepage
               , "qux.com"
@@ -504,8 +512,10 @@ createProjectTest pkgIx srcDb = testGroup "createProject tests"
               -- license
               , "3"
               -- author
+              , "git username"
               , "Foobar"
               -- email
+              , "git email"
               , "foobar@qux.com"
               -- homepage
               , "qux.com"
@@ -579,8 +589,10 @@ createProjectTest pkgIx srcDb = testGroup "createProject tests"
               -- license
               , "3"
               -- author
+              , "git username"
               , "Foobar"
               -- email
+              , "git email"
               , "foobar@qux.com"
               -- homepage
               , "qux.com"
@@ -660,8 +672,10 @@ createProjectTest pkgIx srcDb = testGroup "createProject tests"
               -- license
               , "3"
               -- author
+              , "git username"
               , "Foobar"
               -- email
+              , "git email"
               , "foobar@qux.com"
               -- homepage
               , "qux.com"
@@ -728,7 +742,9 @@ fileCreatorTests pkgIx srcDb _pkgName = testGroup "generators"
               , "y"               -- "yes to prompt internal to package name"
               , "0.2.0.1"         -- package version
               , "2"               -- pick the second license in the list
+              , "git username"    -- name guessed by calling "git config user.name"
               , "Foobar"          -- author name
+              , "git email"       -- email guessed by calling "git config user.email"
               , "foobar@qux.com"  -- maintainer email
               , "qux.com"         -- package homepage
               , "Qux's package"   -- package synopsis
@@ -834,10 +850,14 @@ interactiveTests srcDb = testGroup "Check top level getter functions"
         , testSimplePrompt "2" synopsisPrompt
             "Resistance is futile, you will be assimilated" ["Resistance is futile, you will be assimilated"]
         ]
-    , testSimplePrompt "Check authorPrompt output" authorPrompt
-        "Foobar" ["Foobar"]
-    , testSimplePrompt "Check emailPrompt output" emailPrompt
-        "foobar@qux.com" ["foobar@qux.com"]
+    , testSimplePrompt "Check authorPrompt output (name supplied by the user)" authorPrompt
+        "Foobar" ["git username", "Foobar"]
+    , testSimplePrompt "Check authorPrompt output (name guessed from git config)" authorPrompt
+        "git username" ["git username", ""]
+    , testSimplePrompt "Check emailPrompt output (email supplied by the user)" emailPrompt
+        "foobar@qux.com" ["git email", "foobar@qux.com"]
+    , testSimplePrompt "Check emailPrompt output (email guessed from git config)" emailPrompt
+        "git@email" ["git@email", ""]
     , testSimplePrompt "Check homepagePrompt output" homepagePrompt
         "qux.com" ["qux.com"]
     , testSimplePrompt "Check testDirsPrompt output" testDirsPrompt
-- 
2.25.1

